README.md
# PHP GoogleTranslate free [![Build Status](https://travis-ci.org/SiegSB/php-google-translate-free.svg?branch=master)](https://travis-ci.org/SiegSB/php-google-translate-free)
Simple PHP library for talking to Google's Translate API for free.

Live example here: https://statickidz.com/proyectos/traductor-google-api-php-gratis/


CHANGELOG.md
# GoogleTranslate PHP Library Changelog
Author: Adrián Barrio Andrés
Author e-mail: statickidz@gmail.com
Collaborator: Paris Niarfe Baltazar Salguero 
Collaborator Twitter: @SiegSBDeveloper
Collaborator e-mail: sieg.sb@gmail.com

## v2.0.0, released on 0:50 15/11/2016  
### New  
* Now the GoogleTranslate library it can import using PSR-4 autoload  
* The library is ready to upload to the composer service for much more simple installation  
* GoogleTranslate now has with it necessary to generate automatic documentation, see the observations for more information  
* Now Travis-CI is supported with PHPUnit tests  
* GoogleTranslate now have a GPL-3.0 licence to make more compatible with others licences. See the observations for more information  
### Modified  
* The library now has its namespace so that is more easy of implement, can see the files in the ' tests' folder for examples of use  
* The comments in the code were improved to facilitate the documentation generation, see the observations for more information
### Observations  
* The documentation is generated automatically using [phpDocumentor](https://www.phpdoc.org)  
* I suggest the change of license to GPL because the previous  license enters in conflict with the same library, the PHP licence 3.0 prohibits the use of the word ' PHP' in the name of the library. In addition the License GPL is more compatible with other licenses and can promote the use of the library

## v1.0.0
This is the first and original release of GoogleTranslate Library by Statickidz  


tests/TranslateTest.php
<?php
require_once ('vendor/autoload.php');
use \Statickidz\GoogleTranslate;

class TranslateTest extends PHPUnit_Framework_TestCase
{

    public function testTranslate()
    {
        $source = 'es';
        $target = 'en';
        $text = 'verdadero';

        $trans = new GoogleTranslate();
        $result = $trans->translate($source, $target, $text);

        $this->assertEquals($result, 'true');
    }
}
?>

example.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', 1);

require_once('GoogleTranslate.class.php');

$source = 'es';
$target = 'en';
$text = "'Destiny' tiene un problema con el balance de sus armas y en Bungie lo saben mejor que nadie. Durante el último año, numerosas han sido las actualizaciones que han ido haciendo ajustes a todos los niveles, más pequeños o más grandes, en busca de un equilibrio que es prácticamente imposible, todo ello en su afán de conseguir que todas las familias de armas puedan resultar igual de relevantes.En la práctica, esto se ha traducido muchas veces en desnudar un santo para vestir a otro, haciendo que al final hubiera siempre una familia de armas que se impusiera sobre el resto en cuanto a porcentaje de uso, algo que a Bungie no parece gustarle nada. Con 'El Rey de los Poseídos', eso es cierto, se ha alcanzado un punto medianamente aceptable, al menos si lo comparamos con el periodo de terror que cañones de mano y escopetas habían impuesto en los meses anteriores, pero la situación dista de ser perfecta, y ahora son los fusiles de pulsos quienes copan el mayor protagonismo en 'Destiny'. Diciembre ha sido elegido por sus creadores para lanzar una nueva actualización a gran escala que volverá a cambiar las reglas del juego por completo. ¿Era necesaria? En mi opinión, no especialmente, o al menos no al nivel que se está planteando, pero imagino que son ellos quienes tienen los datos más apropiados y, por tanto, quienes pueden actuar en consecuencia. Está por ver si con ello consiguen equilibrar aún más la complicada balanza entre familias de armas, algo especialmente difícil si tenemos en cuenta además los imposibles paralelismos entre modalidades PvP y PvE, o vuelven a romper la baraja como ya pasó en anteriores ocasiones.";

$translation = GoogleTranslate::translate($source, $target, $text);

echo '<pre>';
print_r($translation);
echo '</pre>';



GoogleTranslate.class.php
<?php

/**
 * GoogleTranslate.class.php
 *
 * Class to talk with Google Translator.
 *
 * @category   Translation
 * @author     Adrián Barrio Andrés
 * @copyright  2016 Adrián Barrio Andrés
 * @license    http://www.php.net/license/3_0.txt  PHP License 3.0
 * @version    1.0
 * @link       https://statickidz.com/
 */

class GoogleTranslate
{

    /**
     * @param string $source
     * @param string $target
     * @param string $text
     * @return string
     */
    public static function translate($source, $target, $text) {

        // Request translation
        $response = self::requestTranslation($source, $target, $text);

        // Get translation text
        //$response = self::getStringBetween("onmouseout=\"this.style.backgroundColor='#fff'\">", "</span></div>", strval($response));

        // Clean translation
        $translation = self::getSentencesFromJSON($response);

        return $translation;
    }

    /**
     * @param string $source
     * @param string $target
     * @param string $text
     * @return array
     */
    protected static function requestTranslation($source, $target, $text) {

        // Google translate URL
        $url = "https://translate.google.com/translate_a/single?client=at&dt=t&dt=ld&dt=qca&dt=rm&dt=bd&dj=1&hl=es-ES&ie=UTF-8&oe=UTF-8&inputm=2&otf=2&iid=1dd3b944-fa62-4b55-b330-74909a99969e";

        $fields = array(
            'sl' => urlencode($source),
            'tl' => urlencode($target),
            'q' => urlencode($text)
        );

        // URL-ify the data for the POST
        $fields_string = "";
        foreach($fields as $key=>$value) {
            $fields_string .= $key.'='.$value.'&';
        }
        
        rtrim($fields_string, '&');

        // Open connection
        $ch = curl_init();

        // Set the url, number of POST vars, POST data
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_POST, count($fields));
        curl_setopt($ch, CURLOPT_POSTFIELDS, $fields_string);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_ENCODING, 'UTF-8');
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
        curl_setopt($ch, CURLOPT_USERAGENT, 'AndroidTranslate/5.3.0.RC02.130475354-53000263 5.1 phone TRANSLATE_OPM5_TEST_1');

        // Execute post
        $result = curl_exec($ch);

        // Close connection
        curl_close($ch);

        return $result;
    }


    /**
     * @param string $json
     * @return string
     */
    protected static function getSentencesFromJSON($json) {
        $sentencesArray = json_decode($json, true);
        $sentences = "";

        foreach ($sentencesArray["sentences"] as $s) {
            $sentences .= $s["trans"];
        }

        return $sentences;
    }

}


src/GoogleTranslate.php
<?php
namespace Statickidz;

/**
 * GoogleTranslate.class.php
 *
 * Class to talk with Google Translator for free.
 *
 * @package PHP Google Translate Free;
 * @category Translation
 * @author Adrián Barrio Andrés
 * @author Paris N. Baltazar Salguero <sieg.sb@gmail.com>
 * @copyright 2016 Adrián Barrio Andrés
 * @license https://opensource.org/licenses/GPL-3.0 GNU General Public License 3.0
 * @version 2.0
 * @link https://statickidz.com/
 */

/**
 * Main class GoogleTranslate
 *
 * @package GoogleTranslate
 *
 */
class GoogleTranslate
{

    /**
     * Retrieves the translation of a text
     *
     * @param string $source
     *            Original language of the text on notation xx. For example: es, en, it, fr...
     * @param string $target
     *            Language to which you want to translate the text in format xx. For example: es, en, it, fr...
     * @param string $text
     *            Text that you want to translate
     *
     * @return string a simple string with the translation of the text in the target language
     */
    public static function translate($source, $target, $text)
    {
        // Request translation
        $response = self::requestTranslation($source, $target, $text);

        // Get translation text
        // $response = self::getStringBetween("onmouseout=\"this.style.backgroundColor='#fff'\">", "</span></div>", strval($response));

        // Clean translation
        $translation = self::getSentencesFromJSON($response);

        return $translation;
    }

    /**
     * Internal function to make the request to the translator service
     *
     * @internal
     *
     * @param string $source
     *            Original language taken from the 'translate' function
     * @param string $target
     *            Target language taken from the ' translate' function
     * @param string $text
     *            Text to translate taken from the 'translate' function
     *
     * @return object[] The response of the translation service in JSON format
     */
    protected static function requestTranslation($source, $target, $text)
    {

        // Google translate URL
        $url = "https://translate.google.com/translate_a/single?client=at&dt=t&dt=ld&dt=qca&dt=rm&dt=bd&dj=1&hl=es-ES&ie=UTF-8&oe=UTF-8&inputm=2&otf=2&iid=1dd3b944-fa62-4b55-b330-74909a99969e";

        $fields = array(
            'sl' => urlencode($source),
            'tl' => urlencode($target),
            'q' => urlencode($text)
        );

        // URL-ify the data for the POST
        $fields_string = "";
        foreach ($fields as $key => $value) {
            $fields_string .= $key . '=' . $value . '&';
        }

        rtrim($fields_string, '&');

        // Open connection
        $ch = curl_init();

        // Set the url, number of POST vars, POST data
        curl_setopt($ch, CURLOPT_URL, $url);
        curl_setopt($ch, CURLOPT_POST, count($fields));
        curl_setopt($ch, CURLOPT_POSTFIELDS, $fields_string);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_ENCODING, 'UTF-8');
        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, false);
        curl_setopt($ch, CURLOPT_USERAGENT, 'AndroidTranslate/5.3.0.RC02.130475354-53000263 5.1 phone TRANSLATE_OPM5_TEST_1');

        // Execute post
        $result = curl_exec($ch);

        // Close connection
        curl_close($ch);

        return $result;
    }

    /**
     * Dump of the JSON's response in an array
     *
     * @param string $json
     *            The JSON object returned by the request function
     *
     * @return string A single string with the translation
     */
    protected static function getSentencesFromJSON($json)
    {
        $sentencesArray = json_decode($json, true);
        $sentences = "";

        foreach ($sentencesArray["sentences"] as $s) {
            $sentences .= $s["trans"];
        }

        return $sentences;
    }
}
